examples/api/lib/material/theme/theme_extension.1..dart

```diff
@@ -0,0 +1,134 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flutter code sample for ThemeExtension
+
+import 'package:flutter/material.dart';
+
+@immutable
+class MyColors implements ThemeExtension {
+  const MyColors({
+    this.blue,
+    this.red,
+  });
+
+  final Color? blue;
+  final Color? red;
+
+  @override
+  MyColors copyWith({Color? red, Color? blue}) {
+    return MyColors(
+      blue: blue ?? this.blue,
+      red: red ?? this.red,
+    );
+  }
+
+  @override
+  MyColors lerp(ThemeExtension? other, double t) {
+    if (other is MyColors) {
+      return MyColors(
+        blue: Color.lerp(blue, other.blue, t),
+        red: Color.lerp(red, other.red, t),
+      );
+    } else {
+      return this;
+    }
+  }
+
+  // Optional
+  @override
+  bool operator ==(Object other) {
+    return other is MyColors &&
+        other.blue == blue &&
+        other.red == red;
+  }
+
+  // Optional
+  @override
+  int get hashCode {
+    return hashList(<Object?>[
+      blue,
+      red,
+    ]);
+  }
+}
+
+extension on ThemeData {
+  MyColors get myColors => themeExtension! as MyColors;
```

Comment:
nit: Have been wondering if there's a type safe way to ensure that the ThemeExtension set in the ThemeData will always be the same type as the ThemeData we're trying to get. But haven't thought of any clean solutions. Would be nice but I don't think it's necessary, feel free to ack (:

Reply:
Maybe 
  MyColors? get myColors => themeExtensio is MyColors ? themeExtension : null;

Reply:
That works, but adding runtime type-checking and nullability here means null checks in code, slight tradeoff.

```dart
Theme.of(context).myColors?.blue
```